### PR TITLE
kernel: mmu: make demand paging work on SMP

### PR DESCRIPTION
**Pull request by npitre**
This is the bare minimum for demand paging to work on SMP systems.

Tested on ARM64 (with extra patches for ARM64 demand paging support).

### CODE REVIEW

**Comment by dcpleung** [comment]
Hm... isn't that mutex has some restrictions on when it can be used (something about being inside interrupt handling)? Since those functions can be called under lots of circumstances, maybe mutex is not the best choice?

**Comment by npitre** [comment]
> Hm... isn't that mutex has some restrictions on when it can be used
> (something about being inside interrupt handling)? Since those functions can
> be called under lots of circumstances, maybe mutex is not the best choice?

My use case has its backing store sitting on top of a filesystem on top of
the flashmap layer on top of the flash layer on top of the SPI layer with DMA
transfer. Any page fault must be ready to sleep. So no, demand-paged
interrupt handlers are pretty much out of the question.

If 2 cores trigger demand paging at the same time, one of them must be able
to schedule something else or go to sleep. These mutexes create a 
serialization point but even with parallel scratch pages things would get 
serialized at the flash driver anyway.

The code for `mutex_lock()` starts with 

```
    __ASSERT(!arch_is_in_isr(), "mutexes cannot be used inside ISRs");
```

which is perfect in those circumstances.

**Comment by dcpleung** [comment]
That sounds good.

**Review by dcpleung — DISMISSED** [review]

File: kernel/mmu.c [hunk 1]
```diff
@@ -1174,14 +1175,16 @@ static inline void do_backing_store_page_out(uintptr_t location)
...
+#if defined(CONFIG_SMP) && defined(CONFIG_DEMAND_PAGING_ALLOW_IRQ)
+/*
+ * SMP support is embryonic at this point. Some resources such as the
+ * scratch page would need to be per CPU, backing store driver execution
+ * would have to be confined to the faulting CPU, statistics would need
+ * to cope with access concurrency, etc. For now let's simply enforce global
+ * serialization across all CPUs with a mutex.
  */
-BUILD_ASSERT(!IS_ENABLED(CONFIG_SMP));
+static struct k_mutex z_mm_paging_lock;
```
Comment (peter-mitsis) [inline kernel/mmu.c:1186]: z_mm_paging_lock appears to be relying upon the BSS properties to initialize it. Please using either k_mutex_init(), or define it using K_MUTEX_DEFINE().

File: kernel/mmu.c [hunk 2]
```diff
@@ -1374,19 +1391,23 @@ int k_mem_page_frame_evict(uintptr_t phys)
 	}
 
 #ifdef CONFIG_DEMAND_PAGING_ALLOW_IRQ
-	irq_unlock(key);
+	k_spin_unlock(&z_mm_lock, key);
```
Comment (peter-mitsis) [inline kernel/mmu.c:1394]: Am I correct in thinking that  CONFIG_DEMAND_PAGING_ALLOW_IRQ can not be used in the SMP case? In the UP case a reschedule is prevented from occurring during the call to do_backing_store_page_out() by either the irq/spin lock or k_sched_lock() depending upon whether CONFIG_DEMAND_PAGING_ALLOW_IRQ is enabled. 

I may be wrong, but I am presently assuming that the reschedule prevention properties are why k_sched_lock() was chosen for UP. However, in the SMP case, the k_mutex_lock() does not prevent any rescheduling on that cpu if CONFIG_DEMAND_PAGING_ALLOW_IRQ is enabled. If SMP and allow IRQ do not mix, then the Kconfig for CONFIG_DEMAND_PAGING_ALLOW_IRQ should be updated.

I have some knowledge gaps here, so I can not say for certain what the desired behavior should be so I am throwing it out there.

**Comment by npitre** [comment]
@peter-mitsis commented:

> Am I correct in thinking that  CONFIG_DEMAND_PAGING_ALLOW_IRQ can not be
> used in the SMP case?
>  
> In the UP case a reschedule is prevented from occurring during the call
> to `do_backing_store_page_out()` by either the irq/spin lock or
> `k_sched_lock()` depending upon whether `CONFIG_DEMAND_PAGING_ALLOW_IRQ`
> is enabled.
> 
> I may be wrong, but I am presently assuming that the reschedule prevention
> properties are why `k_sched_lock()` was chosen for UP. However, in the SMP
> case, the `k_mutex_lock()` does not prevent any rescheduling on that cpu
> if `CONFIG_DEMAND_PAGING_ALLOW_IRQ` is enabled. If SMP and allow IRQ do
> not mix, then the Kconfig for `CONFIG_DEMAND_PAGING_ALLOW_IRQ` should be
> updated.

We must wonder why rescheduling is prevented in the first place.

On UP, you may assume that a cooperative thread will never be replaced by
another thread unless some explicit reschedule-inducing action is performed
by that thread. Being scheduled away because some arbitrary memory access
triggered a demand paging event would break that assumption, hence the
`k_sched_lock()`.

On SMP you may not assume exclusivity solely from being a cooperative thread.
Another thread with any prio may be running on another CPU. Exclusion must
already be enforced by other means. Therefore trying to prevent scheduling on
SMP is pointless. And `k_sched_lock()` is equivalent to a no-op on SMP anyway. 

In any case I do need for demand paging to reschedule and/or sleep. Given my
target is SMP it is OK.

**Comment by npitre** [comment]
Updated the comment in the code to reflect the above rationale.

**Comment by npitre** [comment]
Clarified comments some more.

Please review. This is the last generic patch before ARM64 demand paging
support can go in.

**Review by peter-mitsis — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Pull request merged**